nipype/pipeline/engine/workflows.py

```diff
@@ -1065,88 +1071,425 @@ class Join(Node):
+    @property
+    def global_done(self):
+        # once _global_done os True, this should not change
+        logger.debug('global_done {}'.format(self._global_done))
+        if self._global_done:
+            return self._global_done
+        else:
+            return self._check_all_results()
+
+    # dj: version without join
+    def _check_all_results(self):
+        # checking if all files that should be created are present
+        for ind in itertools.product(*self.state._all_elements):
+            state_dict = self.state.state_values(ind)
+            dir_nm_el = "_".join(["{}.{}".format(i, j) for i, j in list(state_dict.items())])
+            for key_out in self._out_nm:
+                if not os.path.isfile(os.path.join(self.nodedir, dir_nm_el, key_out+".txt")):
+                    return False
+        self._global_done = True
+        return True
+
+
+    # reading results (without join for now)
+    @property
+    def result(self):
+        if not self._result:
+            self._reading_results()
+        return self._result
+
+
+    def _reading_results(self):
+        """
+        reading results from file,
+        doesn't check if everything is ready, i.e. if self.global_done"""
+        for key_out in self._out_nm:
+            self._result[key_out] = []
+            if self._state_inputs:
+                files = [name for name in glob.glob("{}/*/{}.txt".format(self.nodedir, key_out))]
+                for file in files:
+                    st_el = file.split(os.sep)[-2].split("_")
+                    st_dict = collections.OrderedDict([(el.split(".")[0], eval(el.split(".")[1]))
+                                                            for el in st_el])
+                    with open(file) as fout:
+                        logger.debug('Reading Results: file={}, st_dict={}'.format(file, st_dict))
+                        self._result[key_out].append((st_dict, eval(fout.readline())))
+            # for nodes without input
+            else:
+                files = [name for name in glob.glob("{}/{}.txt".format(self.nodedir, key_out))]
+                with open(files[0]) as fout:
+                    self._result[key_out].append(({}, eval(fout.readline())))
+
+
+
+     #
+    # @property
+    # def global_done(self):
+    #     return self.nodecore.global_done
+    #
+    #
+    # @property
+    # def needed_outputs(self):
+    #     return self.nodecore.needed_outputs
+    #
+    # @property
+    # def result(self):
+    #     return self.nodecore.result
+    #
+    #
+    def prepare_state_input(self):
+        self._state.prepare_state_input(state_inputs=self.state_inputs)
+    #     # updating node
+    #     self.nodecore.state = self._state
+    #     self.nodecore._inputs = self._inputs
+    #     self.nodecore._state_inputs = self._state_inputs
+    #
+    def run(self, plugin="serial"):
+        self.prepare_state_input()
+        self.sub = sub.SubmitterNode(plugin, node=self)
```

Comment:
is there a reason to create and attach sub to self? would something like this not work?

```python
def run(self, plugin=None):
    if plugin is None:
        submitter = sub.Submitter(plugin='Serial')
   if isinstance(plugin, Submitter):
        submitter = plugin 
    for parameters in ...:
         jobid = plugin.submit(node.interface, parameters)
         #add jobid to state
```

Reply:
@satra - sure, I can do it like this (although the specific syntax is not clear to me), would just have to provide  `node.interface`, `node.state`, for workflow it will be `wf.state`. `wf.graph` etc. I was thinking about it as another workaround, can always come back to it.

And do you understand why `cf` was complaining?

Reply:
possibly because of recursive memory pointers.